CRM/Core/ManagedEntities.php

```diff
@@ -655,23 +534,41 @@ private function isActivationSupported(string $entity_type): bool {
    * Load declarations into the class property.
    *
    * This picks it up from hooks and enabled components.
-   */
-  protected function loadDeclarations(): void {
-    $this->declarations = [];
-    foreach (CRM_Core_Component::getEnabledComponents() as $component) {
-      $this->declarations = array_merge($this->declarations, $component->getManagedEntities());
+   *
+   * @param string $moduleName
+   *   Limit reconciliation to a single module.
+   */
+  protected function loadDeclarations($moduleName = NULL): void {
+    $declarations = [];
+    // Exclude components if given a module name.
+    if (!$moduleName || $moduleName === 'civicrm') {
+      foreach (CRM_Core_Component::getEnabledComponents() as $component) {
+        $declarations = array_merge($declarations, $component->getManagedEntities());
+      }
     }
-    CRM_Utils_Hook::managed($this->declarations);
-    $this->declarations = $this->cleanDeclarations($this->declarations);
+    // Ideally, given a $moduleName like 'org.foo.demo' we'd just call the function `demo_civicrm_managed()`
+    // But alas, that only works with old-style hooks and not the event dispatcher.
+    // So here we go loading declarations from every module whether we need them or not.
+    CRM_Utils_Hook::managed($declarations);
```

Comment:
Comment responding to a comment... `CRM_Utils_Hook` dispatches to several different systems (Drupal, Joomla, WordPress, Civi extensions, and Symfony listeners). The problem isn't with any one system - it's that multiple-dispatch requires a "lowest common denominator" approach, and several systems (*Symfony+WordPress+Joomla*) don't support the trick you describe. 

If one doesn't need portability or backward-compatibility, then we could define a different interface for registering mgds (and specifically support this).

However, even within the current hook, I think this could be largely improved - eg add a second parameter:

https://github.com/civicrm/civicrm-core/commit/e452a1485a0bf5860bd68370800e9ba6b4064228

This doesn't quite stop the dispatch - but it can prevent unnecessary file-scans (which is probably 90% of the cost). Note that this draft changes the `mixin` -- so the optimization would roll-out to all core extensions (which already use the mixin) and to [future/migrated `civix` extensions](https://github.com/totten/civix/pull/233).

Reply:
@totten I like that commit. I notice that you make it into an array, whereas this PR only considers a single `$module`. We should probably be consistent there. We could cast it to an array as soon as it gets passed into `reconcile()` and then it would always be multivalued.

Reply:
Good idea @colemanw.

I can send up that commit as a follow-up PR.

Reply:
Oops, nevermind ^^^. I see you cherry-picked it. :)